lib/remote_node.go

```diff
@@ -0,0 +1,744 @@
+	ver.Services = rn.nodeServices
+
+	// We use an int64 instead of a uint64 for convenience.
+	ver.TstampSecs = time.Now().Unix()
+
+	ver.Nonce = nonce
+	ver.UserAgent = rn.params.UserAgent
+
+	// When a node asks you for what height you have, you should reply with the height of the latest actual block you
+	// have. This makes it so that peers who have up-to-date headers but missing blocks won't be considered for initial
+	// block download.
+	ver.LatestBlockHeight = rn.latestBlockHeight
+
+	// Set the minimum fee rate the peer will accept.
+	ver.MinFeeRateNanosPerKB = rn.minTxFeeRateNanosPerKB
+
+	return ver
+}
+
+func (rn *RemoteNode) IsTimedOut() bool {
+	if rn.IsTerminated() {
+		return true
+	}
+	if rn.IsConnected() || rn.IsVersionSent() {
+		return rn.versionTimeExpected.Before(time.Now())
+	}
+	if rn.IsVerackSent() {
+		return rn.verackTimeExpected.Before(time.Now())
+	}
+	return false
+}
+
+// HandleVersionMessage is called upon receiving a version message from the RemoteNode's peer. The peer may be the one
+// initiating the handshake, in which case, we should respond with our own version message. To do this, we pass the
+// responseNonce to this function, which we will use in our response version message.
+func (rn *RemoteNode) HandleVersionMessage(verMsg *MsgDeSoVersion, responseNonce uint64) error {
+	rn.mtx.Lock()
+	defer rn.mtx.Unlock()
+
+	if !rn.IsConnected() && !rn.IsVersionSent() {
+		return fmt.Errorf("HandleVersionMessage: RemoteNode is not connected or version exchange has already "+
+			"been completed, connectionStatus: %v", rn.connectionStatus)
+	}
+
+	// Verify that the peer's version matches our minimal supported version.
+	if verMsg.Version < rn.params.MinProtocolVersion {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v) "+
+			"protocol version too low. Peer version: %v, min version: %v", rn.id, verMsg.Version, rn.params.MinProtocolVersion)
+	}
+
+	// Verify that the peer's version message is sent within the version negotiation timeout.
+	if rn.versionTimeExpected.Before(time.Now()) {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v) "+
+			"version timeout. Time expected: %v, now: %v", rn.id, rn.versionTimeExpected.UnixMicro(), time.Now().UnixMicro())
+	}
+
+	vMeta := rn.handshakeMetadata
+	// Record the version the peer is using.
+	vMeta.advertisedProtocolVersion = NewProtocolVersionType(verMsg.Version)
+	// Make sure the latest supported protocol version is ProtocolVersion2.
+	if vMeta.advertisedProtocolVersion.After(ProtocolVersion2) {
+		return fmt.Errorf("RemoteNode.HandleVersionMessage: Requesting disconnect for id: (%v) "+
+			"protocol version too high. Peer version: %v, max version: %v", rn.id, verMsg.Version, ProtocolVersion2)
+	}
+
+	// Decide on the protocol version to use for this connection.
+	negotiatedVersion := rn.params.ProtocolVersion
+	if verMsg.Version < rn.params.ProtocolVersion.ToUint64() {
```

Comment:
I think this check is problematic. If you include it, then you will be partitioning the network between upgraded nodes and non-upgraded nodes. This is likely not what we want. Instead, we want PoS nodes to be able to talk to un-upgraded nodes until at least the first PoS cutoff, and likely the second PoS cutoff as well. After the second cutoff, then it makes more sense to refuse connections to/from un-upgraded nodes.